### PR TITLE
Update and reorganize CONTRIBUTING

### PR DESCRIPTION
**Pull request by micahellison**
It's been a while since CONTRIBUTING was updated and there are a few missing details worth mentioning.

### Changes
* Reference Code of Conduct
* Organizing CONTRIBUTING by different methods of contributing
* Add lots of useful in-line links
* Clarify branching strategy
* Explain ways in which contributors can test
* Explain PR process
* Lots of other small additions/fixes

### Checklist
- [X] The code change is tested and works locally.
- [X] Tests pass. Your PR cannot be merged unless tests pass
- [X] There is no commented out code in this PR.
- [X] Have you followed the guidelines in our Contributing document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [n/a] Have you written new tests for your core changes, as applicable?

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,45 +1,111 @@
...
- * [Developing jrnl](#developing-jrnl)
+## Code of Conduct
 
+Before starting, please read the [Code of Conduct](CODE_OF_CONDUCT.md).
 
-## Docs and Typos
+## Reporting Bugs
 
-If you find a typo or a mistake in the docs, please fix it right away and send a pull request. The Right Way™ to fix the docs is to edit the `docs/*.md` files on the **master** branch. You can see the result if you run `make html` inside the project's root directory, which will open a browser that hot-reloads as you change the docs. This requires [mkdocs](https://www.mkdocs.org) to be installed. The `gh-pages` branch is automatically maintained and updates from `master`; you should never have to edit that.
+Please report bugs by [opening a new issue](https://github.com/jrnl-org/jrnl/issues/new/choose) and describing it as well as possible. Many bugs are specific to a particular operating system and Python version, so details are encouraged.
```
Comment (wren) [inline CONTRIBUTING.md]: The rest of this PR has a great, personal voice, but 'details are encouraged' here is very passive. Can we change to something more personal?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,45 +1,111 @@
...
 
-## Bugs
+## Editing Documentation
 
-Unfortunately, bugs happen. If you found one, please [open a new issue](https://github.com/jrnl-org/jrnl/issues/new/choose) and describe it as well as possible. If you're a programmer with some time, go ahead and send us a pull request that references the issue! We'll review as quickly as we can.
+If you find a typo or a mistake in the docs, please fix it right away and send a pull request.
 
-## Feature requests and ideas
+To edit the documentation, edit the `docs/*.md` files on the **develop** branch. You can see the result if you run `make html` (or `poetry run mkdocs serve` if on Windows) inside the project's root directory, then navigating your browser to [locahost:8000](http://localhost:8000). This requires [mkdocs](https://www.mkdocs.org) to be installed.
```
Comment (wren) [inline CONTRIBUTING.md]: Poetry will install mkdocs, so there's no need to call it out separately here.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,45 +1,111 @@
...
+
+The nature of jrnl means we deal with extremely sensitive data, and can't risk data loss. While jrnl does have a comprehensive automated testing suite, user testing is crucial to mitigating this risk.
+
+### Prereleases
+
+[Prereleases are deployed through PyPi much like normal releases](https://pypi.org/project/jrnl/#history). You can use [pipx](https://pypi.org/project/pipx/) to fetch them and test them. See the [changelog](CHANGELOG.md) for information on what has changed with each release.
+
+### Pull requests
+
+If you are comfortable enough with git, you're encouraged to fetch particular [pull requests](https://github.com/jrnl-org/jrnl/pulls), test them yourself, and report back your findings. Bonus points if you can add a screencast of how the new feature works.
```
Comment (wren) [inline CONTRIBUTING.md]: "You're encouraged" here is more passive than the rest of the PR.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,45 +1,111 @@
...
 * avoid duplicating functionality
 
-## New programmers and programmers new to python
+## Developing
 
-Although jrnl has grown quite a bit since its inception, the overall complexity (for an end-user program) is fairly low, and we hope you'll find the code easy enough to understand.
+### Getting your environment set up
 
-If you have a question, please don't hesitate to ask! Python is known for its welcoming community and openness to novice programmers, so feel free to fork the code and play around with it! If you create something you want to share with us, please create a pull request. We never expect pull requests to be perfect, idiomatic, instantly mergeable code. We can work through it together!
+You will need to install [poetry](https://poetry.eustace.io/) to develop jrnl. It will take care of all of the project's other dependencies, except for [mkdocs](https://www.mkdocs.org), which you must install if you wish to preview documentation changes.
```
Comment (wren) [inline CONTRIBUTING.md]: I'm pretty sure mkdocs isn't a separate requirement. But maybe I'm wrong? Did you test this?

Comment (micahellison) [reply CONTRIBUTING.md]: You're right! I just ensured I didn't have mkdocs installed globally and successfully served the docs. Will edit.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,45 +1,111 @@
...
+When resolving bugs or adding new functionality, please add tests to prevent that functionality from breaking in the future. If you notice any functionality that isn't covered in the tests, feel free to submit a test-only pull request as well.
+
+For testing, jrnl uses [behave](https://behave.readthedocs.io/) tests, which are all in the `features` folder.
+
+Many tests can be created by only editing `feature` files with the same format as other tests. For more complicated functionality, you may need to implement steps in `features/steps` which are then executed by your tests in the `feature` files.
+
+### Submitting pull requests
+
+When you're ready, feel free to submit a pull request. The jrnl maintainers generally review the pull requests every two weeks, but the continuous integration pipeline will run on automated tests on it within a matter of minutes and will report back any issues it has found with your code across a variety of environments.
+
```
Comment (wren) [inline CONTRIBUTING.md]: We should mention that PRs will likely get reviewed if they have failing tests.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,45 +1,111 @@
...
+If you have a feature request or idea for jrnl, please [open a new issue](https://github.com/jrnl-org/jrnl/issues/new/choose) and describe the goal of the feature, and any relevant use cases. We'll discuss the issue with you, and decide if it's a good fit for the project.
 
 When discussing new features, please keep in mind our design goals. jrnl strives to do one thing well. To us, that means:
 
 * be _slim_
 * have a simple interface
 * avoid duplicating functionality
 
-## New programmers and programmers new to python
+## Developing
```
Comment (wren) [inline CONTRIBUTING.md]: Can we add something to this section about looking in the backlog, and what the various labels are for?

Comment (micahellison) [reply CONTRIBUTING.md]: Added a "Finding things to work on" section

**Pull request merged**